gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,396 @@
+    }
+
+    return true;
+  }
+
+  @Override
+  public void renameTable(TableIdentifier from, TableIdentifier to) {
+    String fromDb = databaseId(from.namespace());
+    validateDatabaseId(fromDb, from.namespace());
+
+    String toDb = databaseId(to.namespace());
+    validateDatabaseId(toDb, to.namespace());
+
+    Preconditions.checkArgument(
+        fromDb.equals(toDb),
+        "Cannot rename table %s to %s: database must match",
+        from.toString(),
+        to.toString());
+    client.renameTable(tableName(fromDb, from.name()), tableName(toDb, to.name()));
+  }
+
+  @Override
+  public void createNamespace(Namespace namespace, Map<String, String> metadata) {
+    if (namespace.isEmpty()) {
+      // Used by `CREATE NAMESPACE <catalog>`. Create a BLMS catalog linked with Iceberg catalog.
+      client.createCatalog(catalogName, Catalog.getDefaultInstance());
+      LOG.info("Created BigLake catalog: {}", catalogName.toString());
+    } else if (namespace.levels().length == 1) {
+      // Create a database.
+      String dbId = databaseId(namespace);
+      Database.Builder builder = Database.newBuilder().setType(Database.Type.HIVE);
+      builder
+          .getHiveOptionsBuilder()
+          .putAllParameters(metadata)
+          .setLocationUri(databaseLocation(dbId));
+
+      client.createDatabase(DatabaseName.of(projectId, region, catalogId, dbId), builder.build());
+    } else {
+      throw new IllegalArgumentException(
+          String.format("Invalid namespace (too long): %s", namespace));
+    }
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    if (namespace.isEmpty()) {
+      return Streams.stream(client.listDatabases(catalogName))
+          .map(BigLakeCatalog::namespace)
+          .collect(ImmutableList.toImmutableList());
+    }
+
+    // Database namespace does not have nested namespaces.
+    if (namespace.levels().length == 1) {
+      return ImmutableList.of();
+    }
+
+    throw new NoSuchNamespaceException("Invalid namespace: %s", namespace);
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) {
+    try {
+      if (namespace.isEmpty()) {
+        // Used by `DROP NAMESPACE <catalog>`. Deletes the BLMS catalog linked by Iceberg catalog.
+        client.deleteCatalog(catalogName);
+        LOG.info("Deleted BigLake catalog: {}", catalogName.toString());
+      } else if (namespace.levels().length == 1) {
+        String dbId = databaseId(namespace);
+        validateDatabaseId(dbId, namespace);
+        client.deleteDatabase(databaseName(dbId));
+        // Don't delete the data file folder for safety. It aligns with HMS's default behavior.
+        // To support database or catalog level config controlling file deletion in future.
+      } else {
+        LOG.warn("Invalid namespace (too long): {}", namespace);
+        return false;
+      }
+    } catch (NoSuchNamespaceException e) {
+      LOG.warn("Failed to drop namespace", e);
```

Comment:
This log message is inaccurate. If the database didn't exist, then there was no failure. This method is idempotent.

Reply:
But we can't tell whether it is not found or permission denied. For permission denied, we did want to notice the user to double check. Maybe the BigLake client should not convert 403 to 404, instead just convert 403 to Iceberg NotAuthorizedException? then the downstream never treat these errors as not found.

Reply:
I tried to return NotAuthorizedException from the client instead of 403. The problem is that the tests in CatalogTests expects NoSuchTableException and NoSuchNamespaceException. More refactoring is required. I feel convert 403 to 404 and make it is explicit to users is the best option for now.